src/decode/lzma.rs

```diff
@@ -202,33 +206,33 @@ impl DecoderState {
             is_rep_0long: [0x400; 192],
             state: 0,
             rep: [0; 4],
-            len_decoder: rangecoder::LenDecoder::new(),
-            rep_len_decoder: rangecoder::LenDecoder::new(),
+            len_decoder: LenDecoder::new(),
+            rep_len_decoder: LenDecoder::new(),
         }
     }
 
     pub fn reset_state(&mut self, new_props: LzmaProperties) {
         new_props.validate();
         if self.lzma_props.lc + self.lzma_props.lp == new_props.lc + new_props.lp {
             // We can reset here by filling the existing buffer with 0x400.
-            self.literal_probs.iter_mut().for_each(|v| v.fill(0x400))
+            self.literal_probs.fill(0x400);
         } else {
             // We need to reallocate because of the new size of `lc+lp`.
-            self.literal_probs = vec![vec![0x400; 0x300]; 1 << (new_props.lc + new_props.lp)];
+            self.literal_probs = Vec2D::init(0x400, (1 << (new_props.lc + new_props.lp), 0x300));
         }
 
         self.lzma_props = new_props;
         self.pos_slot_decoder.iter_mut().for_each(|t| t.reset());
         self.align_decoder.reset();
-        self.pos_decoders.fill(0x400);
-        self.is_match.fill(0x400);
-        self.is_rep.fill(0x400);
-        self.is_rep_g0.fill(0x400);
-        self.is_rep_g1.fill(0x400);
-        self.is_rep_g2.fill(0x400);
-        self.is_rep_0long.fill(0x400);
+        self.pos_decoders = [0x400; 115];
+        self.is_match = [0x400; 192];
+        self.is_rep = [0x400; 12];
+        self.is_rep_g0 = [0x400; 12];
+        self.is_rep_g1 = [0x400; 12];
+        self.is_rep_g2 = [0x400; 12];
+        self.is_rep_0long = [0x400; 192];
         self.state = 0;
-        self.rep.fill(0);
+        self.rep = [0; 4];
```

Comment:
I don't think this brings any readability benefit over `fill`. The compiler should be able to optimize both versions of the code regardless.

I see you however observed otherwise:

> I found that it was faster to just recreate stack-allocated arrays in DecoderState than bother with filling them. Heap-backed arrays still use fill to keep their allocation.

Can you add a comment here to explain that?

(IMO, this might also be worth filing a bug to the Rust compiler, to understand if `fill` being slower is intentional or not a missed optimization somewhere in the compiler.)